File: Binary Search Tree/BinarySearchTree.java
class BinarySearchTree{
  public int value;
  public int depth;
  public BinarySearchTree left;
  public BinarySearchTree right;

  public BinarySearchTree(int value,int depth){
    this.value = value;
    this.depth = depth;
    this.right = null;
    this.left = null;
  }

  public BinarySearchTree(int value){
    this.value = value;
    this.depth = 1;
    this.left = null;
    this.right = null;
  }

  public void insert(int value){
    if (value < this.value){
      if(this.left = null){
        this.left = new BinarySearchTree(value,this.depth + 1);
        System.out.printf("Tree node %d added to the %s of %d at depth %d \n", value, "left", this.value, this.depth + 1);
      }else{
        this.left.insert(value);
      }
    }else{
      if(this.right = null){
        this.right = new BinarySearchTree(value,this.depth+1);
        System.out.printf("Tree node %d added to the %s of %d at depth %d \n", value, "right", this.value, this.depth + 1);
      }else{
        this.right.insert(value);
      }
    }
  }

  public BinarySearchTree getNodeByValue(int value){
    if(this.value == null){
      return this;
    }else if (value < this.value && this.left != null){
      return this.left.getNodeByValue(value);
    }else if (value > this.value && this.right != null){
      return this.left.getNodeByValue(value);
    }
    return null;
  }

  public void depthFirstTraversal() {
    if (this.left != null) {
      this.left.depthFirstTraversal();
    }
    System.out.println(this.value);
    if (this.right != null) {
      this.right.depthFirstTraversal();
    }
  }

  public static void main(String[] args) {
    System.out.println("Creating Binary Search Tree rooted at value 15:");
    BinarySearchTree tree = new BinarySearchTree(15);

    for (int i = 0; i < 10; i++) {
      tree.insert((int) (Math.random() * 100));
    }

    System.out.println("Printing the inorder depth-first traversal:");
    tree.depthFirstTraversal();
  }
}

// https://www.codecademy.com/learn/paths/pass-the-technical-interview-with-java/tracks/nonlinear-data-structures-java/modules/binary-search-trees-java/cheatsheet

File: Binary Search Tree/BinarySearchTree.py
import random

class BinarySearchTree:
    def __init__(self, value, depth=1):
        self.value = value
        self.depth = depth
        self.left = None
        self.right = None

    def insert(self, value):
        if value < self.value:
            if self.left is None:
                self.left = BinarySearchTree(value, self.depth + 1)
                print(f"Tree node {value} added to the left of {self.value} at depth {self.depth + 1}")
            else:
                self.left.insert(value)
        else:
            if self.right is None:
                self.right = BinarySearchTree(value, self.depth + 1)
                print(f"Tree node {value} added to the right of {self.value} at depth {self.depth + 1}")
            else:
                self.right.insert(value)

    def get_node_by_value(self, value):
        if self is None:
            return None
        elif value == self.value:
            return self
        elif value < self.value and self.left is not None:
            return self.left.get_node_by_value(value)
        elif value > self.value and self.right is not None:
            return self.right.get_node_by_value(value)
        return None

    def depth_first_traversal(self):
        if self.left is not None:
            self.left.depth_first_traversal()
        print(self.value)
        if self.right is not None:
            self.right.depth_first_traversal()


if __name__ == "__main__":
    print("Creating Binary Search Tree rooted at value 15:")
    tree = BinarySearchTree(15)

    for _ in range(10):
        tree.insert(int(100 * random.random()))

    print("Printing the inorder depth-first traversal:")
    tree.depth_first_traversal()
